

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>QCing and filtering the read alignments &mdash; M. tuberculosis Bioinformatics Workshop 0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Day 4: Phylogenetics/Phylogenomics and Building Reproducible Computational Pipelines" href="../day4/index.html" />
    <link rel="prev" title="Inspecting, summarizing, and manipulating the read alignments" href="mappingstats.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> M. tuberculosis Bioinformatics Workshop
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../agenda.html">Agenda</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../day1/index.html">Day 1: Unix Shell Refresher for Bioinformatics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day2/index.html">Day 2: Microbial Genome Assembly (from short and long reads) and Genome Annotation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Day 3: Read mapping and variant calling</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dataset.html">Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="dataprep_readqc.html">Data Prep and Read QC</a></li>
<li class="toctree-l2"><a class="reference internal" href="mapping.html">Mapping Reads to the Reference Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="igv.html">Visualizing the read alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="mappingstats.html">Inspecting, summarizing, and manipulating the read alignments</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">QCing and filtering the read alignments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#duplicate-marking">Duplicate marking</a></li>
<li class="toctree-l3"><a class="reference internal" href="#assessing-depth-of-coverage">Assessing depth of coverage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#assessing-mapping-qualities">Assessing mapping qualities</a></li>
<li class="toctree-l3"><a class="reference internal" href="#indel-realignment-and-filtering-low-quality-alignments">Indel realignment and filtering low quality alignments</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../day4/index.html">Day 4: Phylogenetics/Phylogenomics and Building Reproducible Computational Pipelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day5/index.html">Day 5: Building Reproducible Computational Pipelines</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">M. tuberculosis Bioinformatics Workshop</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Day 3: Read mapping and variant calling</a> &raquo;</li>
        
      <li>QCing and filtering the read alignments</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/material/day3/alignment_qc_filter.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="qcing-and-filtering-the-read-alignments">
<h1>QCing and filtering the read alignments<a class="headerlink" href="#qcing-and-filtering-the-read-alignments" title="Permalink to this headline">¶</a></h1>
<p>Before we can attempt variant calling, we need to make sure that the alignments that are resulted from mapping artefacts are identified. We also want to get an overall picture on the mapping quality.</p>
<div class="section" id="duplicate-marking">
<h2>Duplicate marking<a class="headerlink" href="#duplicate-marking" title="Permalink to this headline">¶</a></h2>
<p>During the lecture, you heard about library and optical duplicates, which if they exist will bias the variant calling. We will mark these duplicates using <code class="docutils literal notranslate"><span class="pre">picard</span></code>.</p>
<p>First, let’s sort the <code class="docutils literal notranslate"><span class="pre">.bam</span></code> file, this time using <code class="docutils literal notranslate"><span class="pre">picard</span></code>. See the note in the previous section. This should give identical results to <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">sort</span></code>, we are simply doing it this way for pipelining consistency.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>java -Xmx16g -Djava.io.tmpdir=/tmp -jar /opt/picard.jar SortSam \
I=$MYRESULTS/mapping/${sample}.mapped.bam O=$MYRESULTS/mapping/${sample}.sorted.bam SORT_ORDER=coordinate
</pre></div>
</div>
<p>Now, we can mark the duplicates:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>java -Xmx16g -Djava.io.tmpdir=/tmp -jar /opt/picard.jar MarkDuplicates \
I=$MYRESULTS/mapping/${sample}.sorted.bam O=$MYRESULTS/mapping/${sample}.sorted.nodup.bam \
REMOVE_DUPLICATES=true ASSUME_SORT_ORDER=coordinate M=$MYRESULTS/mapping/${sample}.sorted.dupmetrics
</pre></div>
</div>
<p>The metrics file will contain some statistics about the de-duplication. In the resulting <code class="docutils literal notranslate"><span class="pre">bam</span></code> file, only one fragment from each duplicate group survives unchanged, other duplicate fragments are given a FLAG <code class="docutils literal notranslate"><span class="pre">0x400</span></code> and will NOT be used downstream. Optimally, detection and marking of duplicate fragments should be done per library, i.e., over all read groups corresponding to a given library. In practice, often it is fine to take a shortcut and is sufficient to do it per lane (read group) to keep things naively <em>parallelizable</em>.</p>
</div>
<div class="section" id="assessing-depth-of-coverage">
<h2>Assessing depth of coverage<a class="headerlink" href="#assessing-depth-of-coverage" title="Permalink to this headline">¶</a></h2>
<p>You might often want to take a peak at the depth of coverage using the read alignments. You can do this with <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">depth</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>samtools depth -a $MYRESULTS/mapping/${sample}.sorted.nodup.bam &gt; $MYRESULTS/mapping/${sample}.coverage.txt
</pre></div>
</div>
<p>This will give you the depth of coverage for each position in the genome.</p>
<p>Question: How many lines should be in this file?
Question: Can you write an <code class="docutils literal notranslate"><span class="pre">awk</span></code> one-liner to quickly compute the average depth of average?</p>
</div>
<div class="section" id="assessing-mapping-qualities">
<h2>Assessing mapping qualities<a class="headerlink" href="#assessing-mapping-qualities" title="Permalink to this headline">¶</a></h2>
<p>Next, we will generate summary statistics on mapping qualities. For this, we will use <code class="docutils literal notranslate"><span class="pre">qualimap</span></code>.
<code class="docutils literal notranslate"><span class="pre">qualimap</span></code> has a GUI but we don’t want to use it here. To avoid launching the GUI, we do <code class="docutils literal notranslate"><span class="pre">unset</span> <span class="pre">DISPLAY</span></code> first.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>unset DISPLAY;
qualimap bamqc -bam $MYRESULTS/mapping/${sample}.sorted.nodup.bam --java-mem-size=4g -outformat PDF -outdir $MYRESULTS/mapping/${sample}.bam.qualimap
</pre></div>
</div>
<p>The results will be in a text file named ${sample}.bam.qualimap/genome_results.txt. Later when we scale our analysis up to multiple isolates, we will parse out these files to get summary statistics for all of the isolates.</p>
</div>
<div class="section" id="indel-realignment-and-filtering-low-quality-alignments">
<h2>Indel realignment and filtering low quality alignments<a class="headerlink" href="#indel-realignment-and-filtering-low-quality-alignments" title="Permalink to this headline">¶</a></h2>
<p>We will use <code class="docutils literal notranslate"><span class="pre">pilon</span></code> to call our variants. This will give us a candidate list from a single sample, which we can further filter after we merge variants from multiple samples. <code class="docutils literal notranslate"><span class="pre">pilon</span></code> with its <code class="docutils literal notranslate"><span class="pre">--variant</span></code> option will perform <em>indel realignment</em>. We can also specify <em>depth</em> and mapping quality based filters. Run <code class="docutils literal notranslate"><span class="pre">pilon</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>java -Xmx16g -jar /opt/pilon-1.22.jar --genome $MYDATA/ref/NC_000962.3.fna --bam $MYRESULTS/mapping/${sample}.sorted.nodup.bam --output $MYRESULTS/mapping/${sample}_pilon --variant  --mindepth 10 --minmq 40 --minqual 20
</pre></div>
</div>
<p>The output will be a <code class="docutils literal notranslate"><span class="pre">.vcf</span></code> file and a <code class="docutils literal notranslate"><span class="pre">.fasta</span></code> file. <code class="docutils literal notranslate"><span class="pre">.fasta</span></code> file has the sequence from this reference based assembly, which we won’t be using downstream.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../day4/index.html" class="btn btn-neutral float-right" title="Day 4: Phylogenetics/Phylogenomics and Building Reproducible Computational Pipelines" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mappingstats.html" class="btn btn-neutral" title="Inspecting, summarizing, and manipulating the read alignments" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Ulas Karaoz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.1',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  <script type="text/javascript" src="../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>