
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Microbial Genome Annotation &#8212; M. tuberculosis Bioinformatics Workshop 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Day 2" href="../day2/index.html" />
    <link rel="prev" title="Improving the accuracy of the long read assembly" href="longassembly_polish.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../day2/index.html" title="Day 2"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="longassembly_polish.html" title="Improving the accuracy of the long read assembly"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">M. tuberculosis Bioinformatics Workshop 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Day 1</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="microbial-genome-annotation">
<h1>Microbial Genome Annotation<a class="headerlink" href="#microbial-genome-annotation" title="Permalink to this headline">¶</a></h1>
<p>Now that we have assembled the reads into contigs, we will annotate the genome. That means we will predict genes and assign functional annotation to them. <em>M. tuberculosis</em> is a very clonal organism and its genome is already well annotated so typically when we assemble a new <em>M. tuberculosis</em> isolate, there is no need to do <code class="docutils literal notranslate"><span class="pre">de</span> <span class="pre">novo</span></code> annotation. Although we are using <em>M. tuberculosis</em> as a working example here, the methodology is applicable to all microbial genomes.</p>
<p>A set of programs are available for genome annotation tasks. Here we will use <code class="docutils literal notranslate"><span class="pre">Prokka</span></code>, a pipeline comprising several open source bioinformatic tools and databases. <code class="docutils literal notranslate"><span class="pre">Prokka</span></code> automates the process of locating open reading frames (ORFs) and RNA regions on contigs, translating ORFs to protein sequences, searching for protein homologs and producing standard output files. For gene finding and translation, <code class="docutils literal notranslate"><span class="pre">Prokka</span></code> uses a program named <code class="docutils literal notranslate"><span class="pre">prodigal</span></code>. Homology searches are then performed using <code class="docutils literal notranslate"><span class="pre">BLAST</span></code> and <code class="docutils literal notranslate"><span class="pre">HMMER</span></code> based on the translated protein sequences as queries against a set of public databases (CDD, PFAM, TIGRFAM) as well as custom databases that are included with <code class="docutils literal notranslate"><span class="pre">Prokka</span></code>.</p>
<p>Create a directory to hold the annotation results:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="o">~/</span><span class="n">wgs_assembly</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">annotation</span><span class="o">/</span><span class="n">prokka</span>
<span class="n">cd</span> <span class="o">~/</span><span class="n">wgs_assembly</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">annotation</span><span class="o">/</span>
</pre></div>
</div>
<p>Run <code class="docutils literal notranslate"><span class="pre">Prokka</span></code> (will take ~20 minutes for our assembly):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>prokka contigs.fa --outdir $SAMPLE --norrna --notrna --metagenome --cpus 8
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Prokka</span></code> produces several types of output:</p>
<ul class="simple">
<li>a <code class="docutils literal notranslate"><span class="pre">.gff</span></code> file, which is a standardised, tab-delimited, format for genome annotations</li>
<li>a Genbank (<code class="docutils literal notranslate"><span class="pre">.gbk</span></code>) file, which is a more detailed description of nucleotide sequences and the genes encoded in these.</li>
</ul>
<p>When your dataset has been annotated you can view the annotations directly in the <code class="docutils literal notranslate"><span class="pre">gff</span></code> file. Peak into the <code class="docutils literal notranslate"><span class="pre">gff</span></code> file by doing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">more</span> <span class="o">-</span><span class="n">S</span> <span class="n">PROKKA_11242015</span><span class="o">.</span><span class="n">gff</span>
</pre></div>
</div>
<p>Question: How many coding regions were found by Prodigal? (Hint: use <code class="docutils literal notranslate"><span class="pre">grep</span></code>)
Question: How many of the coding regions were given an enzyme identifier?
Question: How many were given a COG identifier?</p>
<div class="section" id="visualizing-the-annotations-with-artemis">
<h2>Visualizing the annotations with Artemis<a class="headerlink" href="#visualizing-the-annotations-with-artemis" title="Permalink to this headline">¶</a></h2>
<p>Artemis is a graphical Java program to browse annotated genomes.
Launch Artemis and load <code class="docutils literal notranslate"><span class="pre">gff</span></code> file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">art</span><span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Microbial Genome Annotation</a><ul>
<li><a class="reference internal" href="#visualizing-the-annotations-with-artemis">Visualizing the annotations with Artemis</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="longassembly_polish.html"
                        title="previous chapter">Improving the accuracy of the long read assembly</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../day2/index.html"
                        title="next chapter">Day 2</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/day1/genome_annotation.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../day2/index.html" title="Day 2"
             >next</a> |</li>
        <li class="right" >
          <a href="longassembly_polish.html" title="Improving the accuracy of the long read assembly"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">M. tuberculosis Bioinformatics Workshop 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Day 1</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, Ulas Karaoz.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.5.
    </div>
  </body>
</html>