

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Improving the accuracy of the long read assembly &mdash; M. tuberculosis Bioinformatics Workshop 0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Microbial Genome Annotation" href="genome_annotation.html" />
    <link rel="prev" title="Assessing the quality of the assemblies" href="assembly_qc.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> M. tuberculosis Bioinformatics Workshop
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../agenda.html">Agenda</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../day1/index.html">Day 1: Unix Shell Refresher for Bioinformatics</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Day 2: Microbial Genome Assembly (from short and long reads) and Genome Annotation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="datasets.html">Data Sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="dataprep.html">Copying data and preparing the directories</a></li>
<li class="toctree-l2"><a class="reference internal" href="read_qc.html">Read Quality Control and Trimming</a></li>
<li class="toctree-l2"><a class="reference internal" href="preassembly_qc.html">Pre-assembly Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="assembly.html">Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="assembly_qc.html">Assessing the quality of the assemblies</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Improving the accuracy of the long read assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="genome_annotation.html">Microbial Genome Annotation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../day3/index.html">Day 3: Read mapping and variant calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day4/index.html">Day 4: Phylogenetics/Phylogenomics and Building Reproducible Computational Pipelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day5/index.html">Day 5: Building Reproducible Computational Pipelines</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">M. tuberculosis Bioinformatics Workshop</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Day 2: Microbial Genome Assembly (from short and long reads) and Genome Annotation</a> &raquo;</li>
        
      <li>Improving the accuracy of the long read assembly</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/material/day2/longassembly_polish.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="improving-the-accuracy-of-the-long-read-assembly">
<h1>Improving the accuracy of the long read assembly<a class="headerlink" href="#improving-the-accuracy-of-the-long-read-assembly" title="Permalink to this headline">Â¶</a></h1>
<p>In the lecture, we have seen that PacBio long reads have a significantly higher error rate compared to Illumina short reads. As a result, the long read assembly will have many errors in the consensus sequence (primarily indels). We can increase the accuracy of the consensus sequences using the Illumina reads, which will provide more coverage and higher quality bases. This process of improving the accuracy of the long reads using high fidelity short reads is sometimes referred to as <em>polishing</em>.</p>
<p>To polish PacBio assembly, we will use <code class="docutils literal notranslate"><span class="pre">pilon</span></code> which will generate a new consensus sequence for the assembly. The details of the following commands will be clear in the next lecture. For now, run the following commands:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>mkdir $MYRESULTSDIR/short_long_reads_assembly/polished_pacbio_assembly
PACBIO_ASSEMBLY=$MYRESULTSDIR/short_long_reads_assembly/assemblies/MTB_pacbio_assembly.fasta
cd $MYRESULTSDIR/short_long_reads_assembly/
bwa index $PACBIO_ASSEMBLY
bwa mem -t 4 -p $PACBIO_ASSEMBLY $MYRESULTSDIR/qc/MTB_illumina60X_qced_R1R2.fastq | samtools sort -T tmp -o $MYRESULTSDIR/short_long_reads_assembly/polished_pacbio_assembly/MTB_illumina60X_qced.mapped_to_pacbio_contigs.sorted.bam -
samtools index $MYRESULTSDIR/short_long_reads_assembly/polished_pacbio_assembly/MTB_illumina60X_qced.mapped_to_pacbio_contigs.sorted.bam
java -Xmx8G -jar /opt/pilon-1.22.jar --genome $PACBIO_ASSEMBLY --frags $MYRESULTSDIR/short_long_reads_assembly/polished_pacbio_assembly/MTB_illumina60X_qced.mapped_to_pacbio_contigs.sorted.bam --output $MYRESULTSDIR/short_long_reads_assembly/polished_pacbio_assembly/MTB_pacbio_assembly_shortreadcorrected
</pre></div>
</div>
<p>Do assembly qc with 4 assemblies. <strong>DONOT RUN</strong>, we will use precomputed results.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">quast</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">R</span> <span class="o">../</span><span class="n">NC_000962</span><span class="o">.</span><span class="mf">3.</span><span class="n">fna</span> <span class="o">-</span><span class="n">o</span> <span class="n">quast_4assemblies_qc</span> <span class="o">./</span><span class="n">MTB_illumina20X_assembly</span><span class="o">.</span><span class="n">fasta</span> <span class="o">./</span><span class="n">MTB_illumina60X_assembly</span><span class="o">.</span><span class="n">fasta</span> <span class="o">./</span><span class="n">MTB_pacbio_assembly</span><span class="o">.</span><span class="n">fasta</span> <span class="o">./</span><span class="n">MTB_pacbio_assembly_shortreadcorrected</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>Copy precomputed results:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cp -r ~/mtb_genomics_workshop_data/genome-assembly-annotation/precalculated/assemblies/quast_4assemblies_qc/ $MYRESULTSDIR/short_long_reads_assembly/assemblies/
</pre></div>
</div>
<p>In <code class="docutils literal notranslate"><span class="pre">quast_4assemblies_qc</span></code> directory, open <code class="docutils literal notranslate"><span class="pre">report.html</span></code> file in a browser. After you open, make sure you turn on the <cite>extended report</cite>.</p>
<dl class="docutils">
<dt>::</dt>
<dd>firefox $MYRESULTSDIR/short_long_reads_assembly/assemblies/quast_4assemblies_qc/report.html&amp;</dd>
</dl>
<p>When you opened the results web page, put your green stickies on. We will discuss these results in detail.</p>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="genome_annotation.html" class="btn btn-neutral float-right" title="Microbial Genome Annotation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="assembly_qc.html" class="btn btn-neutral" title="Assessing the quality of the assemblies" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Ulas Karaoz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.1',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  <script type="text/javascript" src="../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>