

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Processing and Merging SNVs from multiple isolates to get an MSA &mdash; M. tuberculosis Bioinformatics Workshop 0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> M. tuberculosis Bioinformatics Workshop
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../agenda.html">Agenda</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Workshop Computer Labs</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">M. tuberculosis Bioinformatics Workshop</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
      <li>Processing and Merging SNVs from multiple isolates to get an MSA</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/material/day4/merge_snvs_from_vcf.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="processing-and-merging-snvs-from-multiple-isolates-to-get-an-msa">
<h1>Processing and Merging SNVs from multiple isolates to get an MSA<a class="headerlink" href="#processing-and-merging-snvs-from-multiple-isolates-to-get-an-msa" title="Permalink to this headline">¶</a></h1>
<p>What we need to do next is to get a set of high-quality SNVs across the whole genome. A SNV, single nucleotide variant (variant means substitution here), is defined as a variation in a single locus without any reference to its frequency in the population. A SNP or single nucleotide polymorphism, on the other hand is a variation for which we have evidence that it is fixed in the population. That means we should see it in more than a certain (arbitrarily defined but often &gt;1%) number of individuals/isolates. Therefore, not all the SNVs are SNPs. Although this is the working definition we will be following, there is no consensus on this defition and all single nucleotide variants are called SNPs independent of their frequency in the population.</p>
<div class="section" id="previously-in-the-workshop">
<h2>Previously in the workshop…<a class="headerlink" href="#previously-in-the-workshop" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><code class="docutils literal notranslate"><span class="pre">vcf</span></code> file we generated using  <code class="docutils literal notranslate"><span class="pre">pilon</span></code> contains unfiltered SNVs as well as indels. To remind, <code class="docutils literal notranslate"><span class="pre">pilon</span></code> did a series of filtering and quality control steps before coming up with this list SNVs and indels. These steps include</dt>
<dd><ol class="first last arabic simple">
<li>indel realignment</li>
<li>base quality score filtering</li>
<li>mapping quality score filtering</li>
<li>depth of coverage filtering.</li>
</ol>
</dd>
</dl>
<p>So all SNVs in our <code class="docutils literal notranslate"><span class="pre">vcf</span></code> file have been, to the best of our ability, stripped off any known read mapping/alignment artifacts, has a reasonably high base quality, depth of coverage, and mapping quality.</p>
<p>Even after these steps, we expect a relatively high false positive rate within this list so we will need to do some further filtering. The assumption is that we are leaning towards minimizing false positive rate. We might be missing some true variants during this strict filtering steps. However, realize that we want to make sure that what we have as a final list (for instance to infer phylogeny) contain only true variants. That is we don’t care so much if we do miss few true variants here and there (that is higher false negative rate is somewhat acceptable).</p>
</div>
<div class="section" id="why-are-we-doing-all-this">
<h2>Why are we doing all this?<a class="headerlink" href="#why-are-we-doing-all-this" title="Permalink to this headline">¶</a></h2>
<p>The end point for this analysis is to get an MSA (multiple sequence alignment) of the variable sites, which we can then use to infer phylogeny. Keep that in mind as you dive in.</p>
</div>
<div class="section" id="separating-snvs-and-indels-and-filtering-snvs-within-highly-repetitive-regions">
<h2>Separating SNVs and indels and filtering SNVs within highly repetitive regions<a class="headerlink" href="#separating-snvs-and-indels-and-filtering-snvs-within-highly-repetitive-regions" title="Permalink to this headline">¶</a></h2>
<p>First, we will need to get rid of the indels (insertion/deletion). As you’ve heard during the lecture, read mappers donot do a particularly good job with indels. Any called indel by any mapper, you should be very suspicious of. We want to remove them for phylogenetic analysis -we should already have lots of variation to infer phylogeny- as they will bias the phylogenetic inference. In the next step, we will use <code class="docutils literal notranslate"><span class="pre">vcftools</span></code> to make 2 separate <code class="docutils literal notranslate"><span class="pre">vcf</span></code> files, one that contains only SNVs and the other indels. Concurrently, we will also remove any variant that sits within a highly repetitive region. Next, a few words on why we want to do this, particularly in analyzing <em>M. tuberculosis</em> genomes.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Repetitive regions in <em>M. tuberculosis</em> genome: A significant portion of the genome of <em>Mycobacterium tuberculosis</em> contains highly repetitive regions (every microbial genome will have some level of tandem and more complex repetitive regions, regions with low sequence complexity). Short-read sequencing isn’t sufficient to make accurate SNV calls within these regions. Therefore we want to remove these regions from further analysis as they will confuse downstream phylogenetic analysis. We have precalculated the repetitive regions, which is provided in the <code class="docutils literal notranslate"><span class="pre">Mtb_repeats.bed</span></code> file in <code class="docutils literal notranslate"><span class="pre">bed</span></code> format. <a class="reference external" href="https://uswest.ensembl.org/info/website/upload/bed.html">bed</a> is a file format used to specify genomic intervals.
Overall, the total length of repetitive regions specified in <code class="docutils literal notranslate"><span class="pre">Mtb_repeats.bed</span></code> is ~572 Kb (~13% of the genome). The majority of these regions (~10% of the genome) belong to two large unrelaed gene families (PE and PPE families) clustered in the genome. They code for glycine-rich proteins that are often based on multiple copies of the polymorphic repetitive sequences referred to as PGRSs <strong>r</strong>e<strong>s</strong>tructured <strong>t</strong>ext. (<a href="#id1"><span class="problematic" id="id2">**</span></a>P**olymorphic guanine cytosine–rich (GC-rich) repetitive sequences). Despite their prominence in the genome, we almost know nothing about the biological significance of these regions.</p>
</div>
<p>Run <code class="docutils literal notranslate"><span class="pre">vcftools</span></code> to separate SNVs/indels and filter repetitive regions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># get rid of &quot;_pilon&quot; in the filename
mv ${sample}_pilon.vcf ${sample}.vcf
vcftools --gzvcf ${sample}.vcf --exclude-bed $SRCBASE/data/ref/Mtb_repeats.bed --remove-indels --recode --recode-INFO-all --out $sample
vcftools --gzvcf ${sample}.vcf --exclude-bed $SRCBASE/data/ref/Mtb_repeats.bed --keep-only-indels --recode --recode-INFO-all --out ${sample}.indels

# take a peek at the indels file! We will be leaving it behind from this point on
more ${sample}.indels.recode.vcf
# and the vcftools log file
more ${sample}.indels.log
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">vcf files are typically huge (~0.5 GB even for a single microbial genome). Compressing vcf files with <a class="reference external" href="http://www.htslib.org/doc/bgzip.html">bgzip</a>  and indexing them with <a class="reference external" href="http://www.htslib.org/doc/tabix.html">tabix</a>  is the standard way vcf files are stored. Once that is done, further processing of the binary files is done with <code class="docutils literal notranslate"><span class="pre">bcftools</span></code>.</p>
</div>
<p>Compress and index the vcf file. Notice that we won’t use the vcf file for indels for this analysis.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>bgzip ${sample}.recode.vcf
tabix -p vcf ${sample}.recode.vcf.gz
</pre></div>
</div>
<p>We need to change the sample header (which is “SAMPLE”, the very last column within the data fields) in the vcf file to specify that these variants are from <code class="docutils literal notranslate"><span class="pre">$sample</span></code>. This is to ensure data provenance: we will soon merge vcf files from multiple samples and they all have the same header at this point. You can do this with <code class="docutils literal notranslate"><span class="pre">bcftools</span></code> which requires the header name to be stored in a text file, so we write the sample name into a file called “sample” first.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>echo $sample &gt; sample
bcftools reheader -s sample ${sample}.recode.vcf.gz &gt; m.${sample}.recode.vcf.gz # we add this &quot;m&quot; just to make the filename different in this temp file
mv -f m.${sample}.recode.vcf.gz ${sample}.recode.vcf.gz

# reindex
tabix -f -p vcf ${sample}.recode.vcf.gz

# make sure nothing that ``pilon`` didn&#39;t let it &quot;PASS&quot; doesn&#39;t get in
bcftools view --include &#39;FILTER=&quot;PASS&quot;&#39; ${sample}.recode.vcf.gz -O z -o temp

mv -f temp ${sample}.recode.vcf.gz
# reindex and we are done
tabix -f -p vcf ${sample}.recode.vcf.gz
</pre></div>
</div>
<p>All we did here was modifying sample “header” and some simple “data massaging”. If you want, you can unzip the <code class="docutils literal notranslate"><span class="pre">${sample}.recode.vcf.gz</span></code> file, peek into it, and do some other quick sanity check:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># OPTIONAL
gunzip ${sample}.recode.vcf.gz
# pay attention to the column name for the very last column, does it look ok?
more ${sample}.recode.vcf
wc ${sample}.vcf
wc ${sample}.recode.vcf
</pre></div>
</div>
</div>
<div class="section" id="merge-vcf-files-from-multiple-isolates">
<h2>Merge <code class="docutils literal notranslate"><span class="pre">vcf</span></code> files from multiple isolates<a class="headerlink" href="#merge-vcf-files-from-multiple-isolates" title="Permalink to this headline">¶</a></h2>
<p>Assuming that we did read mapping etc. for multiple isolates (which you will learn how to do using <code class="docutils literal notranslate"><span class="pre">snakemake</span></code>), at this point we can merge all the vcf files from multiple samples into a single vcf file. We will use that single <code class="docutils literal notranslate"><span class="pre">vcf</span></code> to define <a href="#id3"><span class="problematic" id="id4">*</span></a>SNP*s that be used for phylogenetic analysis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># the way this following command is constructed is intentionally explicit so you can see what is happening</span>
<span class="c1"># notice that this list of 99 isolates contains our prototyping isolate, &quot;ERS050945&quot;, and 98 additional ones</span>
<span class="n">bcftools</span> <span class="n">merge</span> \
<span class="n">ERS050923</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050924</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050925</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050926</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050927</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050928</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050929</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050930</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050931</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050932</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050933</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050934</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050935</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050936</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050937</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050938</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050939</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050940</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050942</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050943</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050944</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050945</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050946</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050947</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050948</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050949</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050950</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050951</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050953</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050954</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050955</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050956</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050957</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050958</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050959</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050960</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050961</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050962</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050963</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050964</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050965</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050966</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050968</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050969</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050970</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050971</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS050972</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053603</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053604</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053605</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053606</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053607</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053608</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053609</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053610</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053611</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053612</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053613</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053614</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053615</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053616</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053617</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053618</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053619</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053620</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053621</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053622</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053623</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053624</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053625</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053626</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053627</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053628</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053629</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053630</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053631</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053632</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053633</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053634</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053635</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053636</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053637</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053638</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053639</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053641</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053642</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053643</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053644</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053651</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053656</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053659</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053661</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053663</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053667</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053672</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS053684</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS096323</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS181351</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">ERS181603</span><span class="o">.</span><span class="n">recode</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
<span class="o">-</span><span class="n">O</span> <span class="n">v</span> <span class="o">-</span><span class="n">o</span> <span class="n">merged</span><span class="o">-</span><span class="mi">99</span><span class="n">RussianIsolates</span><span class="o">.</span><span class="n">vcf</span>
</pre></div>
</div>
<p>Merging step requires a reasonable amount of memory and takes some time. We will be using precalculated file. Now, what we essentially have is a vcf file that contains all the information to make an MSA of the variable position in the genome. Since everything is already
defined with respect to a reference genome, we don’t need to build an MSA, per se, it will already just fall off once we get rid of the non variable sites. That is what <code class="docutils literal notranslate"><span class="pre">snp-sites</span></code> does.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Merging multiple vcf files, the way it has been implemented here, is highly compute heavy. It requires high memory instances when you have 100s to 1000s of isolates.</p>
</div>
<p>In what follows, we first convert vcf to a fasta file using <code class="docutils literal notranslate"><span class="pre">vcf2phylip.py</span></code> (despite its name, it also does this) before calling SNPs. <strong>compute heavy, DONOT RUN</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># -m: keep all SNVs independent of their frequency, typically with a larger number of isolates</span>
<span class="c1"># we would want to impose a threshold</span>
<span class="n">vcf2phylip</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">fasta</span> <span class="o">--</span><span class="n">phylip</span><span class="o">-</span><span class="n">disable</span> <span class="o">-</span><span class="n">m</span> <span class="mi">1</span> <span class="o">--</span><span class="nb">input</span> <span class="n">merged</span><span class="o">-</span><span class="mi">99</span><span class="n">RussianIsolates</span><span class="o">.</span><span class="n">vcf</span>

<span class="c1"># -m: output fasta, -v: output VCF, -p: output PHYLIP</span>
<span class="n">snp</span><span class="o">-</span><span class="n">sites</span> <span class="o">-</span><span class="n">m</span> <span class="n">merged</span><span class="o">-</span><span class="mi">99</span><span class="n">RussianIsolates</span><span class="o">.</span><span class="n">min1</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="mi">99</span><span class="n">RussianIsolates</span><span class="o">.</span><span class="n">msa</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">99RussianIsolates.msa.fasta</span></code> contains our multiple sequence alignment.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Ulas Karaoz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.1',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  <script type="text/javascript" src="../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>