

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Day 1: Introduction &mdash; M. tuberculosis Bioinformatics Workshop 0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> M. tuberculosis Bioinformatics Workshop
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../agenda.html">Agenda and Lectures’ Slides</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../day1/index.html">Day 1: Unix Shell Refresher for Bioinformatics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day2/index.html">Day 2: Microbial Genome Assembly (from short and long reads) and Genome Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day3/index.html">Day 3: Read mapping and variant calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html">Day 4: Phylogenetics/Phylogenomics and Building Reproducible Computational Pipelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day5/index.html">Day 5: Building Reproducible Computational Pipelines</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">M. tuberculosis Bioinformatics Workshop</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
      <li>Day 1: Introduction</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/material/day4/snakemake.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="day-1-introduction">
<h1>Day 1: Introduction<a class="headerlink" href="#day-1-introduction" title="Permalink to this headline">¶</a></h1>
<p>Today we will give an introduction to sequencing techniques and data processing with linux.</p>
<div class="section" id="preprocessing-quality-filtering-using-a-snakemake-workflow">
<h2>Preprocessing &amp; quality filtering using a Snakemake workflow<a class="headerlink" href="#preprocessing-quality-filtering-using-a-snakemake-workflow" title="Permalink to this headline">¶</a></h2>
<p>In this part the following topics will be covered</p>
<ul class="simple">
<li>introduction to Snakemake</li>
<li>unzipping files</li>
<li>making FASTQC plots</li>
<li>trimming reads with low quality and adapter removal</li>
<li>combining all steps into a single workflow</li>
</ul>
</div>
<div class="section" id="introduction-to-snakemake">
<h2>Introduction to Snakemake<a class="headerlink" href="#introduction-to-snakemake" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://speakerdeck.com/johanneskoester/workflow-management-with-snakemake">https://speakerdeck.com/johanneskoester/workflow-management-with-snakemake</a></p>
<ul class="simple">
<li>Documentation for snakemake can be found here: <a class="reference external" href="https://bitbucket.org/johanneskoester/snakemake/wiki/Documentation">https://bitbucket.org/johanneskoester/snakemake/wiki/Documentation</a></li>
<li>Use this as a reference while following the step-by-step guide below</li>
</ul>
</div>
<div class="section" id="launch-snakemake">
<h2>Launch Snakemake<a class="headerlink" href="#launch-snakemake" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p class="first">Boot up the Ubuntu live USB-stick</p>
</li>
<li><p class="first">Open a terminal window using the icon on the left on the screen or by typing: <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">alt</span> <span class="pre">+</span> <span class="pre">T</span></code></p>
</li>
<li><p class="first">Start the Snakemake environment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">source</span> <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">virtualenvwrapper</span><span class="o">.</span><span class="n">sh</span> <span class="o">&amp;&amp;</span> <span class="n">workon</span> <span class="n">snakemake</span>
</pre></div>
</div>
</li>
<li><p class="first">Make sure the snakemake executable is available:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><strong>Question</strong>: Which version of Snakemake is installed?</p>
</div>
</div>
<div class="section" id="create-a-rule-to-unzip-fastq-files">
<h2>Create a rule to unzip FASTQ files<a class="headerlink" href="#create-a-rule-to-unzip-fastq-files" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p class="first">Create a directory for this part for the workshop:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">qc_snakemake</span>
<span class="n">cd</span> <span class="n">qc_snakemake</span>
</pre></div>
</div>
</li>
</ul>
<p>In Snakemake, workflows are specified as Snakefiles. Inspired by GNU Make, a Snakefile contains rules, that denote how to create output files from input files. Dependencies between rules are handled implicitly, by matching filenames of input files against output files. Thereby wildcards can be used to write general rules.</p>
<ul>
<li><p class="first">Create the main workflow file using your favorite text editor. For a graphical editor Gedit is recommended and for a command-line editor Nano is a good choice.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gedit</span> <span class="n">Snakefile</span> <span class="o">&amp;</span>
</pre></div>
</div>
</li>
<li><p class="first">Add the following lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">configfile</span><span class="p">:</span> <span class="s2">&quot;config.json&quot;</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="s2">&quot;bunzip/</span><span class="si">{sample}</span><span class="s2">.fastq&quot;</span><span class="p">,</span><span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">])</span>

<span class="n">rule</span> <span class="n">bunzip</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">expand</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{basedir}{data}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">basedir</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;basedir&quot;</span><span class="p">],</span><span class="n">data</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">data</span><span class="p">])</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;bunzip/</span><span class="si">{data}</span><span class="s2">.fastq&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;bunzip/</span><span class="si">{data}</span><span class="s2">.log&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="n">shell</span><span class="p">:</span> <span class="s2">&quot;bunzip2 -d -c </span><span class="si">{input}</span><span class="s2"> &gt; </span><span class="si">{output}</span><span class="s2">&quot;</span>
</pre></div>
</div>
</li>
<li><p class="first">Save the file. If you are using Gedit use the File menu or type <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">s</span></code> or in Nano type <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code></p>
</li>
</ul>
<p>In the above <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> first the configuration file is loaded, next a target rule is defined that will trigger the creation of the required output. Finally a rule is made to unzip FASTQ files compressed in bzip2 format.</p>
<p>Since version 3.1, Snakemake directly supports the configuration of your workflow. A configuration is provided as a JSON file. The JSON file can be used to define a dictionary of configuration parameters and their values. In the workflow, the configuration is accessible via the global variable <code class="docutils literal notranslate"><span class="pre">config</span></code>.</p>
<ul>
<li><p class="first">Create a config file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gedit</span> <span class="n">config</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
</li>
<li><p class="first">Add the following configuration:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s2">&quot;basedir&quot;</span><span class="p">:</span> <span class="s2">&quot;/home/nmp/Documents/DeHollander/&quot;</span><span class="p">,</span>
    <span class="s2">&quot;samples&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sample1&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sample2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Sample3&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sample3&quot;</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sample1.fastq.bz2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sample2.fastq.bz2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Sample3&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sample3.fastq.bz2&quot;</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="s2">&quot;adapters&quot;</span><span class="p">:</span> <span class="s2">&quot;/home/nmp/Documents/DeHollander/contaminant_list.txt&quot;</span><span class="p">,</span>
    <span class="s2">&quot;adapters_fasta&quot;</span><span class="p">:</span> <span class="s2">&quot;/home/nmp/Documents/DeHollander/illumina_truseq_adapters.fa&quot;</span>
<span class="p">}</span>
</pre></div>
</div>
</li>
<li><p class="first">Now run snakemake. It will unpack the input files using the <code class="docutils literal notranslate"><span class="pre">bunzip2</span></code> command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span>
</pre></div>
</div>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<dl class="last docutils">
<dt><strong>Questions</strong>:</dt>
<dd><ul class="first last simple">
<li>Can you explain what happened? Which files are used for input? Which output files are created?</li>
<li>Open one of the fastq files either using the File Browser or using <code class="docutils literal notranslate"><span class="pre">less</span></code> on the command line</li>
</ul>
</dd>
</dl>
</div>
</div>
<div class="section" id="creating-quality-plots-using-fastqc">
<h2>Creating quality plots using FASTQC<a class="headerlink" href="#creating-quality-plots-using-fastqc" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p class="first">Add the following text rule to the Snakefile after the bunzip rule:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">fastqc_raw</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span> <span class="n">fastq</span><span class="o">=</span><span class="s2">&quot;bunzip/</span><span class="si">{sample}</span><span class="s2">.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span> <span class="s2">&quot;fastqc_raw/</span><span class="si">{sample}</span><span class="s2">_fastqc/&quot;</span>
    <span class="n">params</span><span class="p">:</span> <span class="nb">dir</span><span class="o">=</span><span class="s2">&quot;fastqc_raw&quot;</span><span class="p">,</span> <span class="n">adapters</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;adapters&#39;</span><span class="p">]</span>
    <span class="n">log</span><span class="p">:</span> <span class="s2">&quot;fastqc_raw.log&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="n">shell</span><span class="p">:</span> <span class="s2">&quot;fastqc -q -t </span><span class="si">{threads}</span><span class="s2"> --contaminants </span><span class="si">{params.adapters}</span><span class="s2"> --outdir </span><span class="si">{params.dir}</span><span class="s2"> </span><span class="si">{input.fastq}</span><span class="s2"> &gt; </span><span class="si">{params.dir}</span><span class="s2">/</span><span class="si">{log}</span><span class="s2">&quot;</span>
</pre></div>
</div>
</li>
<li><p class="first">Add final fastqc outputs to the target rule <code class="docutils literal notranslate"><span class="pre">all</span></code>. Is should now look like this (note that a comma is added at the end of the bunzip line):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="s2">&quot;bunzip/</span><span class="si">{sample}</span><span class="s2">.fastq&quot;</span><span class="p">,</span><span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">]),</span>
        <span class="n">expand</span><span class="p">(</span><span class="s2">&quot;fastqc_raw/</span><span class="si">{sample}</span><span class="s2">_fastqc/&quot;</span><span class="p">,</span><span class="n">sample</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">])</span>
</pre></div>
</div>
</li>
<li><p class="first">Make the plots by running Snakemake again:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span>
</pre></div>
</div>
</li>
<li><p class="first">Have a look at the results using a browser:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">firefox</span> <span class="n">fastqc_raw</span><span class="o">/</span><span class="n">Sample</span><span class="o">*</span><span class="n">_fastqc</span><span class="o">/</span><span class="n">fastqc_report</span><span class="o">.</span><span class="n">html</span>
</pre></div>
</div>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<dl class="last docutils">
<dt><strong>Questions</strong>:</dt>
<dd><ul class="first last simple">
<li>What are the difference between the samples?</li>
<li>Is there any difference in quality?</li>
<li>Can you find adapter contamination?</li>
</ul>
</dd>
</dl>
</div>
</div>
<div class="section" id="trim-low-quality-reads-and-remove-adapters">
<h2>Trim low quality reads and remove adapters<a class="headerlink" href="#trim-low-quality-reads-and-remove-adapters" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p class="first">Create a rule for the fastq-mcf program:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">trim</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span> <span class="n">fastq</span><span class="o">=</span><span class="s2">&quot;bunzip/</span><span class="si">{sample}</span><span class="s2">.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span> <span class="n">fastq</span><span class="o">=</span><span class="s2">&quot;fastq-mcf/</span><span class="si">{sample}</span><span class="s2">_trimmed.fastq&quot;</span>
    <span class="n">params</span><span class="p">:</span> <span class="nb">dir</span><span class="o">=</span><span class="s2">&quot;fastq-mcf&quot;</span><span class="p">,</span> <span class="n">adapters</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;adapters_fasta&quot;</span><span class="p">]</span>
    <span class="n">log</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">_fastq-mcf.log&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="c1"># -C: number of subsamples used for determining adapter parameters</span>
    <span class="c1"># http://scotthandley.wordpress.com/2013/09/25/adapter-removal-and-adaptive-quality-trimming-with-parity-using-fastq-mcf/</span>
    <span class="n">shell</span><span class="p">:</span> <span class="s2">&quot;fastq-mcf </span><span class="si">{params.adapters}</span><span class="s2"> </span><span class="si">{input.fastq}</span><span class="s2"> -o </span><span class="si">{output.fastq}</span><span class="s2"> -C 99999 -P 33 -w 4 -q 25 -u -x 0.01 -l 20 -t 0 -S &gt;&gt; </span><span class="si">{params.dir}</span><span class="s2">/</span><span class="si">{log}</span><span class="s2">&quot;</span>
</pre></div>
</div>
</li>
<li><p class="first">Add it to the Snakefile after the fastqc-raw rule</p>
</li>
<li><p class="first">Run snakemake now with a parameter showing the command that is going to be performed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">-</span><span class="n">p</span>
</pre></div>
</div>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<dl class="last docutils">
<dt><strong>Exercise</strong>:</dt>
<dd><ul class="first last simple">
<li>Have a look at the fastq-mcf parameters. What is the minimum quality.</li>
<li>What do the other parameters mean?</li>
</ul>
</dd>
</dl>
</div>
</div>
<div class="section" id="create-quality-plots-of-the-trimmed-data">
<h2>Create quality plots of the trimmed data<a class="headerlink" href="#create-quality-plots-of-the-trimmed-data" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<dl class="last docutils">
<dt><strong>Exercise</strong>:</dt>
<dd><ul class="first last simple">
<li>Create a rule for creating plots with FASTQC on the trimmed data.</li>
<li>Use the fastqc_rule as as a template and adjust the input and output parameters.</li>
<li>Add the fastqc_trim output to the <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">all:</span></code> by adding another expand() line with the output of the fastqc_trim rule as input.</li>
<li>Run snakemake</li>
<li>Open the output files in Firefox and compare them to the raw data. What is the effect of trimming the reads?</li>
</ul>
</dd>
</dl>
</div>
</div>
<div class="section" id="create-quality-plots-using-r">
<h2>Create quality plots using R<a class="headerlink" href="#create-quality-plots-using-r" title="Permalink to this headline">¶</a></h2>
<p>Is it possible to run R code directly from a Snakefile. To make similar quality plots as FASTQC in R we are going to make use of the ‘qrqc’ package: <a class="reference external" href="http://www.bioconductor.org/packages/release/bioc/html/qrqc.html">http://www.bioconductor.org/packages/release/bioc/html/qrqc.html</a></p>
<ul>
<li><p class="first">First we need to install some required software on the USB. Copy paste the commands below to perform this installation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">bitbucket</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">johanneskoester</span><span class="o">/</span><span class="n">snakemake</span><span class="o">/</span><span class="n">raw</span><span class="o">/</span><span class="n">ef669ba49226368fc71eb1975a89168dcb87d59a</span><span class="o">/</span><span class="n">snakemake</span><span class="o">/</span><span class="n">utils</span><span class="o">.</span><span class="n">py</span>
<span class="n">mv</span> <span class="n">utils</span><span class="o">.</span><span class="n">py</span> <span class="o">~/.</span><span class="n">virtualenvs</span><span class="o">/</span><span class="n">snakemake</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="o">.</span><span class="mi">3</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">snakemake</span><span class="o">/</span>
<span class="n">sudo</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="o">-</span><span class="n">y</span> <span class="n">install</span> <span class="n">python3</span><span class="o">-</span><span class="n">dev</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">rpy2</span>
</pre></div>
</div>
</li>
<li><p class="first">Create a rule that reads in the untrimmed and trimmed data and saves a quality plot to disk:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">snakemake.utils</span> <span class="k">import</span> <span class="n">R</span>

<span class="n">rule</span> <span class="n">qc_plot</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span> <span class="n">untrimmed</span><span class="o">=</span><span class="s2">&quot;bunzip/</span><span class="si">{sample}</span><span class="s2">.fastq&quot;</span><span class="p">,</span> <span class="n">trimmed</span><span class="o">=</span><span class="s2">&quot;fastq-mcf/</span><span class="si">{sample}</span><span class="s2">_trimmed.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span> <span class="s2">&quot;qrqc/</span><span class="si">{sample}</span><span class="s2">_qualplot.png&quot;</span>
    <span class="n">run</span><span class="p">:</span>
        <span class="n">R</span><span class="p">(</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        library(&quot;qrqc&quot;)</span>

<span class="s2">        s1 &lt;- readSeqFile(&quot;</span><span class="si">{input.untrimmed}</span><span class="s2">&quot;)</span>
<span class="s2">        s2 &lt;- readSeqFile(&quot;</span><span class="si">{input.trimmed}</span><span class="s2">&quot;)</span>
<span class="s2">        q &lt;- qualPlot(list(untrimmed = s1, trimmed = s2))</span>
<span class="s2">        ggsave(&quot;</span><span class="si">{output}</span><span class="s2">&quot;, q)</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">Add this rule again to the Snakefile.</p>
</li>
<li><p class="first">Adjust the <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">all:</span></code> to incude the output of the qc_plot rule.</p>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><strong>Exercise</strong>: Extend above rule to the other functions of the qrqc package. Have a look at the documentation of the qrqc package to see what is possible. For example create a base frequency plot with the basePlot command.</p>
</div>
</div>
<div class="section" id="merge-paired-end-illumina-data">
<h2>Merge paired-end Illumina data<a class="headerlink" href="#merge-paired-end-illumina-data" title="Permalink to this headline">¶</a></h2>
<p>Sequence that are produced by the Illumina HiSeq or MiSeq machines come in pairs of fastq files with each pair of reads. Files containing R1 in the filename are in the forward direction and with R2 in the reverse. The two read pairs can be merged into a single, longer sequence. Downstream analysis can benefit of the longer read length. Several tools excist to merge paired-end read, and in this section we will make use of FLASH (<a class="reference external" href="http://ccb.jhu.edu/software/FLASH/">http://ccb.jhu.edu/software/FLASH/</a>).</p>
<ul>
<li><p class="first">First we need to setup a new directory to store the results:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">nmp</span><span class="o">/</span><span class="n">mergepairs</span> <span class="o">&amp;&amp;</span> <span class="n">cd</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">nmp</span><span class="o">/</span><span class="n">mergepairs</span>
</pre></div>
</div>
</li>
<li><p class="first">Create a new config.json file with this content:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s2">&quot;basedir&quot;</span><span class="p">:</span> <span class="s2">&quot;/home/nmp/Documents/DeHollander/&quot;</span><span class="p">,</span>
    <span class="s2">&quot;samples&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Sample4&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sample4&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Sample5&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sample5&quot;</span><span class="p">]</span>
    <span class="p">},</span>
    <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Sample4&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;forward&quot;</span><span class="p">:</span> <span class="s2">&quot;sample4_R1.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;reverse&quot;</span><span class="p">:</span> <span class="s2">&quot;sample4_R2.fastq&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Sample5&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;forward&quot;</span><span class="p">:</span> <span class="s2">&quot;sample5_1.fq&quot;</span><span class="p">,</span> <span class="s2">&quot;reverse&quot;</span><span class="p">:</span> <span class="s2">&quot;sample5_2.fq&quot;</span><span class="p">}</span>
    <span class="p">},</span>
    <span class="s2">&quot;adapters_fasta&quot;</span><span class="p">:</span> <span class="s2">&quot;/home/nmp/Documents/DeHollander/illumina_truseq_adapters.fa&quot;</span>
<span class="p">}</span>
</pre></div>
</div>
</li>
<li><p class="first">The first step again is to trim the data. Since in the previous section we have worked with single end data, we have to adjust the fastq-mcf rule:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">fastqmcf_paired</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">forward</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;basedir&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">data</span><span class="p">][</span><span class="s1">&#39;forward&#39;</span><span class="p">],</span>
        <span class="n">reverse</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;basedir&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">][</span><span class="n">wildcards</span><span class="o">.</span><span class="n">data</span><span class="p">][</span><span class="s1">&#39;reverse&#39;</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">forward</span><span class="o">=</span><span class="s2">&quot;fastq-mcf/</span><span class="si">{data}</span><span class="s2">_R1_trimmed.fastq&quot;</span><span class="p">,</span>
        <span class="n">reverse</span><span class="o">=</span><span class="s2">&quot;fastq-mcf/</span><span class="si">{data}</span><span class="s2">_R2_trimmed.fastq&quot;</span>
    <span class="n">params</span><span class="p">:</span> <span class="nb">dir</span><span class="o">=</span><span class="s2">&quot;fastq-mcf&quot;</span><span class="p">,</span> <span class="n">adapters</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;adapters_fasta&quot;</span><span class="p">]</span>
    <span class="n">log</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{data}</span><span class="s2">_fastq-mcf.log&quot;</span>
    <span class="n">threads</span><span class="p">:</span> <span class="mi">1</span>
    <span class="c1"># -C: number of subsamples used for determining adapter parameters</span>
    <span class="c1"># http://scotthandley.wordpress.com/2013/09/25/adapter-removal-and-adaptive-quality-trimming-with-parity-using-fastq-mcf/</span>
    <span class="n">shell</span><span class="p">:</span> <span class="s2">&quot;fastq-mcf </span><span class="si">{params.adapters}</span><span class="s2"> </span><span class="si">{input}</span><span class="s2"> -o </span><span class="si">{output.forward}</span><span class="s2"> -o </span><span class="si">{output.reverse}</span><span class="s2"> -C 99999 -P 33 -w 4 -q 25 -u -x 0.01 -l 20 -t 0 -S &gt;&gt; </span><span class="si">{params.dir}</span><span class="s2">/</span><span class="si">{log}</span><span class="s2">&quot;</span>
</pre></div>
</div>
</li>
<li><p class="first">We need to get a list of samples from the config file. Add this line to the Snakefile after the config.json file is loaded:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DATA_TO_SAMPLE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">data</span><span class="p">:</span> <span class="n">sample</span> <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">datasets</span> <span class="ow">in</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">datasets</span><span class="p">}</span>
</pre></div>
</div>
</li>
<li><p class="first">Now lets run Snakemake to create a trimmed file of the forward read from Sample4:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="n">fastq</span><span class="o">-</span><span class="n">mcf</span><span class="o">/</span><span class="n">Sample4_R1_trimmed</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
</li>
</ul>
<p>One of the strengths of Snakemake is that you can run it on multiple input files with a range of parameters automatically. We are going to run the FLASH program on Sample4 and Sample5 with a range of mismatch parameters. Therefore we need to create a MISMATCHES variable and a ‘final’ rule that will trigger the creation of all needed files.</p>
<ul>
<li><p class="first">Add these lines directly after the <code class="docutils literal notranslate"><span class="pre">DATA_TO_SAMPLE</span></code> variable:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MISMATCHES</span> <span class="o">=</span> <span class="s2">&quot;0 0.1 0.5&quot;</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="s2">&quot;flash/</span><span class="si">{data}</span><span class="s2">_trimmed_</span><span class="si">{mismatch}</span><span class="s2">.extendedFrags.fastq&quot;</span><span class="o">.</span><span class="n">split</span><span class="p">(),</span><span class="n">data</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">],</span><span class="n">mismatch</span><span class="o">=</span><span class="n">MISMATCHES</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">Now add a rule to run FLASH on a range of mismatch values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="n">flash_trimmed</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">forward</span><span class="o">=</span><span class="s2">&quot;fastq-mcf/</span><span class="si">{data}</span><span class="s2">_R1_trimmed.fastq&quot;</span><span class="p">,</span>
        <span class="n">reverse</span><span class="o">=</span><span class="s2">&quot;fastq-mcf/</span><span class="si">{data}</span><span class="s2">_R2_trimmed.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="s2">&quot;flash/</span><span class="si">{data}</span><span class="s2">_trimmed_</span><span class="si">{mismatch}</span><span class="s2">.extendedFrags.fastq&quot;</span>
    <span class="n">params</span><span class="p">:</span>
        <span class="n">sample</span><span class="o">=</span><span class="k">lambda</span> <span class="n">wildcards</span><span class="p">:</span> <span class="n">DATA_TO_SAMPLE</span><span class="p">[</span><span class="n">wildcards</span><span class="o">.</span><span class="n">data</span><span class="p">],</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{data}</span><span class="s2">_trimmed_</span><span class="si">{mismatch}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">dir</span><span class="o">=</span><span class="s2">&quot;flash&quot;</span><span class="p">,</span><span class="n">minlength</span><span class="o">=</span><span class="s2">&quot;5&quot;</span><span class="p">,</span><span class="n">maxlength</span><span class="o">=</span><span class="s2">&quot;400&quot;</span><span class="p">,</span><span class="n">mismatch</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{mismatch}</span><span class="s2">&quot;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="s2">&quot;</span><span class="si">{data}</span><span class="s2">_trimmed_</span><span class="si">{mismatch}</span><span class="s2">_mergepairs.log&quot;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s2">&quot;flash -m </span><span class="si">{params.minlength}</span><span class="s2"> -M </span><span class="si">{params.maxlength}</span><span class="s2"> </span><span class="si">{input}</span><span class="s2"> -x </span><span class="si">{params.mismatch}</span><span class="s2"> -o </span><span class="si">{params.prefix}</span><span class="s2"> -d </span><span class="si">{params.dir}</span><span class="s2"> &gt; </span><span class="si">{params.dir}</span><span class="s2">/</span><span class="si">{log}</span><span class="s2">&quot;</span>
</pre></div>
</div>
</li>
<li><p class="first">To visualize the steps Snakemake is going to perform on the different datasets, we can create a dependency graph with the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ snakemake --dag | dot | display
</pre></div>
</div>
</li>
<li><p class="first">Finally, let’s  run Snakemake to create all of the files seen in the previous picture:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span>
</pre></div>
</div>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<dl class="last docutils">
<dt><strong>Questions</strong>:</dt>
<dd><ul class="first last simple">
<li>Have a look at the fastq-mcf log file <code class="docutils literal notranslate"><span class="pre">fastq-mcf/Sample4_fastq-mcf.log</span></code>. Do you see any adapter contamination? Do the same for Sample5.</li>
<li>How many reads are merged? Compare the FLASH log file for Sample4 and Sample5: <code class="docutils literal notranslate"><span class="pre">flash/Sample4_trimmed_0_mergepairs.log</span></code></li>
<li>What is the effect of the different mismatch values?</li>
</ul>
</dd>
</dl>
</div>
<p>The main purpose of merging read pairs is to get longer reads. Using the qrqc package in <code class="docutils literal notranslate"><span class="pre">R</span></code> we can compare the length of the sequences before and after merging.</p>
<ul>
<li><p class="first">Create the following <code class="docutils literal notranslate"><span class="pre">qrqc_seqlen</span></code> rule at the end of Snakefile:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">snakemake.utils</span> <span class="k">import</span> <span class="n">R</span>

<span class="n">rule</span> <span class="n">qrqc_seqlen</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span> <span class="n">untrimmed</span><span class="o">=</span><span class="s2">&quot;fastq-mcf/</span><span class="si">{data}</span><span class="s2">_R1_trimmed.fastq&quot;</span><span class="p">,</span> <span class="n">trimmed</span><span class="o">=</span><span class="s2">&quot;flash/</span><span class="si">{data}</span><span class="s2">_trimmed_0.extendedFrags.fastq&quot;</span>
    <span class="n">output</span><span class="p">:</span> <span class="s2">&quot;qrqc/</span><span class="si">{data}</span><span class="s2">_lengthplot.png&quot;</span>
    <span class="n">run</span><span class="p">:</span>
        <span class="n">R</span><span class="p">(</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        library(&quot;qrqc&quot;)</span>

<span class="s2">        s1 &lt;- readSeqFile(&quot;</span><span class="si">{input.untrimmed}</span><span class="s2">&quot;)</span>
<span class="s2">        s2 &lt;- readSeqFile(&quot;</span><span class="si">{input.trimmed}</span><span class="s2">&quot;)</span>
<span class="s2">        l &lt;- seqlenPlot(list(untrimmed = s1, trimmed = s2))</span>
<span class="s2">        ggsave(&quot;</span><span class="si">{output}</span><span class="s2">&quot;, l)</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p class="first">Create the plot for Sample4 and have a look at it:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="n">qrqc</span><span class="o">/</span><span class="n">Sample4_lengthplot</span><span class="o">.</span><span class="n">png</span>
<span class="n">display</span> <span class="n">qrqc</span><span class="o">/</span><span class="n">Sample4_lengthplot</span><span class="o">.</span><span class="n">png</span>
</pre></div>
</div>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<dl class="last docutils">
<dt><strong>Optional exercise</strong>:</dt>
<dd><ul class="first last simple">
<li>Make a rule to merge data with FLASH using the untrimmed files</li>
<li>There are alternative programs to perform the trimming and merging. Create rules for Alientrimmer to trim the data and Usearch to merge data <code class="docutils literal notranslate"><span class="pre">usearch</span> <span class="pre">-fastq_mergepairs</span></code>. Compare this with the output of Fastq-mcf and FLASH.</li>
</ul>
</dd>
</dl>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Ulas Karaoz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.1',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  <script type="text/javascript" src="../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>